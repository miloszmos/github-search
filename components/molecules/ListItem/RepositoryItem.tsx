import { formatDistanceStrict } from 'date-fns';
import React from 'react';
import { Repository } from '../../../graphql/generated/graphql';
import RepositoryIcon from '../../atoms/icons/RepositoryIcon';
import StarIcon from '../../atoms/icons/StarIcon';
import Dot from '../../atoms/Label/DotLabel';
import IconCountLabel from '../../atoms/Label/IconCountLabel';
import { LabelStyles } from '../../atoms/Label/Label.styles';
import {
  InfoStyles,
  ItemTextStyles,
  ItemWrapperStyles,
  LabelWrapper,
} from './ListItem.styles';

interface RepositoryItemProps {
  repository: Repository;
}

const RepositoryItem = ({ repository }: RepositoryItemProps) => {
  const {
    nameWithOwner,
    url,
    description,
    stargazerCount,
    licenseInfo,
    updatedAt,
    languages,
  } = repository;
  const lang = languages?.nodes;

  return (
    <ItemWrapperStyles>
      <div>
        <RepositoryIcon />
      </div>
      <div className="right">
        <a href={url} target="_blank" rel="noreferrer">
          <ItemTextStyles>{nameWithOwner}</ItemTextStyles>
        </a>
        <ItemTextStyles className="gray desc">{description}</ItemTextStyles>
        <InfoStyles>
          <LabelWrapper>
            <IconCountLabel icon={<StarIcon />} count={stargazerCount} />
          </LabelWrapper>
          {lang && lang[0] && (
            <LabelWrapper>
              <Dot label={lang[0].name} bgColor={lang[0].color} />
            </LabelWrapper>
          )}
          {licenseInfo && <LabelWrapper>{licenseInfo.name}</LabelWrapper>}
          <LabelStyles>
            Updated{' '}
            {formatDistanceStrict(new Date(updatedAt), new Date(), {
              addSuffix: true,
            })}
          </LabelStyles>
        </InfoStyles>
      </div>
    </ItemWrapperStyles>
  );
};

export default RepositoryItem;
