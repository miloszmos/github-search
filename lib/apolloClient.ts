import { ApolloClient, InMemoryCache, createHttpLink } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';

const httpLink = createHttpLink({
  uri: process.env.NEXT_PUBLIC_API_ENDPOINT,
});

const authLink = setContext((_, { headers }) => ({
  headers: {
    ...headers,
    authorization: `Bearer ${process.env.NEXT_PUBLIC_ACCESS_TOKEN}`,
  },
}));

export const client = new ApolloClient({
  link: authLink.concat(httpLink),
  cache: new InMemoryCache(),
});
