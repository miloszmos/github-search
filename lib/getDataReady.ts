import { User, Repository, SearchQuery } from '../graphql/generated/graphql';

type ResponseDataType = (Repository | User | null)[];

export const getDataReady = (data: SearchQuery): ResponseDataType => {
  const repoNodes = data.repositories.nodes ?? [];
  const userNodes = data.users.nodes ?? [];
  const mergedData = [...repoNodes, ...userNodes].filter(Boolean) as (
    | Repository
    | User
  )[];

  return mergedData
    .filter(
      // ? Check for User | Repository because I had some other types in response
      (el) => el.__typename && el.__typename.match(/^(User|Repository)$/)
    )
    .sort((a, b) => a.id.localeCompare(b.id));
};
