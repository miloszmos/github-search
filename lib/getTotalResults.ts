import { SearchQuery } from '../graphql/generated/graphql';

export const getTotalResults = (data: SearchQuery): number => {
  return data.repositories.repositoryCount + data.users.userCount;
};
