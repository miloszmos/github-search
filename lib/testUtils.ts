import casual from 'casual';

casual.seed(777);

const fakeUser = () => ({
  __typename: 'User',
  name: casual.name,
  avatarUrl: casual.url,
  bio: casual.description,
  id: casual.uuid,
  location: casual.address,
  login: casual.username,
});

const fakeRepository = () => ({
  __typename: 'Repository',
  id: casual.uuid,
  nameWithOwner: casual.full_name,
  createdAt: '2022-12-11T20:00:00.797Z',
  description: casual.description,
  stargazerCount: casual.random,
  url: casual.url,
  updatedAt: '2022-12-11T20:00:00.797Z',
  licenseInfo: {
    name: casual.name,
  },
  languages: [
    {
      nodes: {
        name: casual.name,
        color: casual.color_name,
      },
    },
  ],
});

export { fakeUser, fakeRepository };
