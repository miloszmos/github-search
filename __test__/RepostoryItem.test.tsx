import React from 'react';
import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import { fakeRepository } from '../lib/testUtils';
import RepositoryItem from '../components/molecules/ListItem/RepositoryItem';

const item = fakeRepository();

describe('<RepositoryItem />', () => {
  it('renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <RepositoryItem repository={item} />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders out base information', async () => {
    const { container } = render(
      <MockedProvider>
        <RepositoryItem repository={item} />
      </MockedProvider>
    );
    const nameWithOwner = screen.getByText(item.nameWithOwner);
    expect(nameWithOwner).toBeInTheDocument();
    await screen.findByTestId('updatedAt');

    const link = container.querySelector('a');
    expect(link).toHaveAttribute('href', 'http://www.Lilly.info/');
    expect(link).toHaveTextContent('Marisa Ward');
  });
});
