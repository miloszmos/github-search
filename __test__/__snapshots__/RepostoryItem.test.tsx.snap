// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RepositoryItem /> renders and matches snapshot 1`] = `
<div>
  <div
    class="sc-b08aa592-2 kghpTT"
  >
    <div>
      <svg
        fill="none"
        height="18"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.167 17.325H13c1.005 0 2.5-.666 2.5-2.5V3.16c0-1.835-1.495-2.5-2.5-2.5H.5v1.666h12.49c.385.01.843.163.843.834 0 .084-.007.159-.02.227-.093.48-.486.598-.822.606H1.333c-.015 0-.025.007-.04.008H.5V15.66c0 .919.748 1.666 1.667 1.666Zm0-11.666h11.666v9.166c0 .672-.458.824-.833.834H7.167V9.825L5.5 10.66l-1.667-.834v5.834H2.167v-10Z"
          fill="#6F7781"
        />
      </svg>
    </div>
    <div
      class="right"
    >
      <a
        href="http://www.Lilly.info/"
        rel="noreferrer"
        target="_blank"
      >
        <p
          class="sc-b08aa592-0 wvFcc"
        >
          Marisa Ward
        </p>
      </a>
      <p
        class="sc-b08aa592-0 wvFcc gray desc"
      >
        Enim quia autem et modi tenetur amet modi reprehenderit. Adipisci eius laborum eius aut illo dolorum. Aliquam id sit. Vel ipsa tempora aut sint aperiam occaecati illum animi vel.
      </p>
      <div
        class="sc-b08aa592-4 cdqKdz"
      >
        <span
          class="sc-d9d70f80-0 sc-b08aa592-3 crKzRe aqaio"
        >
          <div
            class="sc-c5a584b3-0 loTkHG"
          >
            <span
              class="icon"
            >
              <svg
                fill="none"
                height="14"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m8 2.26 1.38 2.79.23.5.5.075 3.08.445L11 8.22l-.375.365.09.5.525 3.065-2.755-1.445L8 10.5l-.465.245L4.78 12.17l.5-3.065.09-.5L5 8.22 2.79 6.045 5.87 5.6l.5-.075.23-.5L8 2.26ZM8 0 5.725 4.61.64 5.345l3.68 3.59L3.45 14 8 11.61 12.55 14l-.87-5.065 3.68-3.585-5.085-.74L8 0Z"
                  fill="#6F7781"
                />
              </svg>
            </span>
            <span
              class="sc-d9d70f80-0 sc-c5a584b3-2 crKzRe haifYZ"
            >
              0.5326204823795706
            </span>
          </div>
        </span>
        <span
          class="sc-d9d70f80-0 sc-b08aa592-3 crKzRe aqaio"
        >
          Dr. Okey Lang
        </span>
        <span
          class="sc-d9d70f80-0 crKzRe"
          data-testid="updatedAt"
        >
          Updated
           
          in 4 months
        </span>
      </div>
    </div>
  </div>
</div>
`;
